### PR TITLE
Remove memory-debug stuff; use asan, other compiler tools

### PR DESCRIPTION
**Pull request by richsalz**
This also speeds up memory allocation, avoiding an atomic increment for each {malloc,realloc,free}

## Checklist
Thank you for your contribution. Please answer the following:

- [X] I acknowledge that I am authorized to submit this code under
the terms of the [Apache License](https://www.apache.org/licenses/LICENSE-2.0)
- [ ] All new public APIs are documented
- [ ] All new public functions have tests
- [ ] All new public functions have fuzzing tests

### CODE REVIEW

**Comment by richsalz** [comment]
Sigh, I always had high hopes for the memory-debug-fail stuff.  Never happened.

**Review by botovq — COMMENTED** [review]
I'm generally fine with this particular removal. Note that there are still a few callers of `CRYPTO_mem_ctrl()` out there. In the OpenBSD ports tree we have audio/mumble, audio/umurmur and telephony/resiprocate that still "use" it. I'm not sure how up-to-date these ports are and if later versions have adapted.

I think `OPENSSL_NO_CRYPTO_MDEBUG` should  be defined in configuration.h.in

In general, I don't think you can rely on the ecosystem having adapted to the latest round of deprecations yet. I expect that it will take a few more years, and there's the other aspect that the alternative to the deprecated APIs often are only available in OpenSSL 3 since very little of this deprecation was actually planned thoroughly with application authors in mind.

Removing some deprecated API will make it hard for an application to target quictls, OpenSSL 3, BoringSSL and LibreSSL at the same time since the latter two don't have OpenSSL 3 API yet (some of it might come at some point, but much of it is just out of the question). I suspect same goes for aws-lc but I haven't checked recently.

**Comment by richsalz** [comment]
Thanks for the info about deprecation.  I think that, if we do anything, we have to stick to 0.9.8 abd 1.x deprecations.

Added the requested define.

File: crypto/mem.c [hunk 1]
```diff
@@ -290,72 +209,4 @@ void CRYPTO_clear_free(void *str, size_t num, const char *file, int line)
         OPENSSL_cleanse(str, num);
     CRYPTO_free(str, file, line);
 }
-
-#if !defined(OPENSSL_NO_CRYPTO_MDEBUG)
-
-# ifndef OPENSSL_NO_DEPRECATED_3_0
```
Comment (botovq) [inline crypto/mem.c:296]: I agree that for now removing Only pre-3.0 deprecations is the right approach. These functions were only deprecated in 3.0, though.

Comment (richsalz) [reply crypto/mem.c:296]: agreed, do not remove things deprecated in 3.0 or later.

**Comment by richsalz** [comment]
Did a force-push that only removes crypto-mdebug,crypto-mdebug-backtrace and does not remove 3.0 deprecations.

File: crypto/mem.c [hunk 2]
```diff
@@ -291,7 +162,6 @@ void CRYPTO_clear_free(void *str, size_t num, const char *file, int line)
     CRYPTO_free(str, file, line);
 }
 
-#if !defined(OPENSSL_NO_CRYPTO_MDEBUG)
 
 # ifndef OPENSSL_NO_DEPRECATED_3_0
```
Comment (botovq) [inline crypto/mem.c:166]:
```suggestion
#ifndef OPENSSL_NO_DEPRECATED_3_0
```

File: crypto/mem.c [hunk 3]
```diff
@@ -358,4 +228,3 @@ int CRYPTO_mem_leaks_cb(int (*cb)(const char *str, size_t len, void *u),
 
 # endif
```
Comment (botovq) [inline crypto/mem.c:229]:
```suggestion
#endif
```

File: include/openssl/configuration.h.in [hunk 4]
```diff
@@ -79,7 +79,9 @@ extern "C" {
 # ifndef OPENSSL_NO_EGD
 #  define OPENSSL_NO_EGD
 # endif
-
+# ifndef CRYPTO_NO_MDEBUG
+#  define CRYPTO_NO_MDEBUG
+# endif
 #ifndef OPENSSL_NO_QUIC
 #define OPENSSL_NO_QUIC
 #endif
```
Comment (botovq) [inline include/openssl/configuration.h.in:87]: Maybe make this incantation match the others?
```C
# ifndef OPENSSL_NO_QUIC
#  define OPENSSL_NO_QUIC
# endif
```

**Review by botovq — COMMENTED** [review]
Looks good now. Only a few nits.

File: include/openssl/configuration.h.in [hunk 5]
```diff
@@ -79,8 +79,8 @@ extern "C" {
 # ifndef OPENSSL_NO_EGD
 #  define OPENSSL_NO_EGD
 # endif
-# ifndef CRYPTO_NO_MDEBUG
-#  define CRYPTO_NO_MDEBUG
+# ifndef OPENSSL_NO_MDEBUG
+#  define OPENSSL_NO_MDEBUG
```
Comment (botovq) [inline include/openssl/configuration.h.in:83]: I missed that this was wrong. It still is. It should be `OPENSSL_NO_CRYPTO_MDEBUG`. My nit was about the missing indent in the quic stanza below.

Comment (richsalz) [reply include/openssl/configuration.h.in]: I will not do commits while on the phone.  I will not do commits while on the phone... sigh.  sorry for wasting your time.  fixup pushed.

**Review by botovq — APPROVED** [review]

**Pull request merged**